backend/src/main/java/com/google/sps/datastore/ParticipantDatastore.java

```diff
@@ -0,0 +1,118 @@
+  private static final String PROPERTY_MATCHID = "matchId";
+  private static final String PROPERTY_TIMESTAMP = "timestamp";
+
+  /** Datastore */
+  private final DatastoreService datastore;
+
+  /** Constructor that takes in DatastoreService */
+  public ParticipantDatastore(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  /**
+   * Put participant in datastore. Overwrite participant entity if participant with same username
+   * already exists.
+   */
+  public void addParticipant(Participant participant) {
+    // Set properties of entity based on participant fields
+    Entity participantEntity = new Entity(KIND_PARTICIPANT, participant.getUsername());
+    participantEntity.setProperty(PROPERTY_USERNAME, participant.getUsername());
+    participantEntity.setProperty(PROPERTY_STARTTIMEAVAILABLE, participant.getStartTimeAvailable());
+    participantEntity.setProperty(PROPERTY_ENDTIMEAVAILABLE, participant.getEndTimeAvailable());
+    participantEntity.setProperty(PROPERTY_DURATION, participant.getDuration());
+    participantEntity.setProperty(PROPERTY_MATCHID, participant.getMatchId());
+    participantEntity.setProperty(PROPERTY_TIMESTAMP, participant.getTimestamp());
+
+    // Insert entity into datastore
+    datastore.put(participantEntity);
+  }
+
+  /** Return Participant Entity from username, or null if entity is not found */
+  @Nullable
+  private Entity getEntity(String username) {
+    Key participantKey = KeyFactory.createKey(KIND_PARTICIPANT, username);
+    try {
+      return datastore.get(participantKey);
+    } catch (EntityNotFoundException e) {
+      return null;
+    }
+  }
+
+  /** Return Participant from username, or null if participant with username not in datastore */
+  @Nullable
+  public Participant getParticipantFromUsername(String username) {
+    Entity entity = getEntity(username);
+    if (entity == null) {
+      return null;
+    }
+    return getParticipantFromEntity(entity);
+  }
+
+  /** Return participant object from datastore participant entity, or null if entity is null */
+  @Nullable
+  private static Participant getParticipantFromEntity(@Nonnull Entity entity) {
+    // Get entity properties
+    String username = (String) entity.getProperty(PROPERTY_USERNAME);
+    long startTimeAvailable = (long) entity.getProperty(PROPERTY_STARTTIMEAVAILABLE);
+    long endTimeAvailable = (long) entity.getProperty(PROPERTY_ENDTIMEAVAILABLE);
+    int duration = ((Long) entity.getProperty(PROPERTY_DURATION)).intValue();
+    long matchId = (long) entity.getProperty(PROPERTY_MATCHID);
+    long timestamp = (long) entity.getProperty(PROPERTY_TIMESTAMP);
+
+    // Create and return new Participant
+    return new Participant(
+        username, startTimeAvailable, endTimeAvailable, duration, matchId, timestamp);
+  }
+
+  /** Return list of all unmatched participants with same duration */
+  public List<Participant> getSameDurationParticipants(int duration) {
```

Comment:
By this part of the code we've lost the concept of "same-ness" i.e. this code doesn't know about the participant we're comparing to. How about getParticipantsWithDuration (and updated javadocs)?

Reply:
Done.